aiida/restapi/common/identifiers.py

```diff
@@ -0,0 +1,281 @@
+# -*- coding: utf-8 -*-
+"""Utility functions to work with node "full types" which are unique node identifiers."""
+from __future__ import absolute_import
+
+import collections
+import six
+
+FULL_TYPE_CONCATENATOR = '|'
+LIKE_OPERATOR_CHARACTER = '%'
+
+
+def validate_full_type(full_type):
+    """Validate that the `full_type` is a valid full type unique node identifier.
+
+    :param full_type: a `Node` full type
+    :raises ValueError: if the `full_type` is invalid
+    :raises TypeError: if the `full_type` is not a string type
+    """
+    from aiida.common.lang import type_check
+
+    type_check(full_type, six.string_types)
+
+    if FULL_TYPE_CONCATENATOR not in full_type:
+        raise ValueError(
+            'full type `{}` does not include the required concatenator symbol `{}`.'.format(
+                full_type, FULL_TYPE_CONCATENATOR
+            )
+        )
+
+
+def construct_full_type(node_type, process_type):
+    """Return the full type, which uniquely identifies any `Node` with the given `node_type` and `process_type`.
+
+    :param node_type: the `node_type` of the `Node`
+    :param process_type: the `process_type` of the `Node`
+    :return: the full type, which is a unique identifier
+    """
+    if node_type is None:
+        process_type = ''
+
+    if process_type is None:
+        process_type = ''
+
+    return '{}{}{}'.format(node_type, FULL_TYPE_CONCATENATOR, process_type)
+
+
+def get_full_type_filters(full_type):
+    """Return the `QueryBuilder` filters that will return all `Nodes` identified by the given `full_type`.
+
+    :param full_type: the `full_type` unique node identifier
+    :return: dictionary of filters to be passed for the `filters` keyword in `QueryBuilder.append`
+    :raises ValueError: if the `full_type` is invalid
+    :raises TypeError: if the `full_type` is not a string type
+    """
+    validate_full_type(full_type)
+
+    filters = {}
+    node_type, process_type = full_type.split(FULL_TYPE_CONCATENATOR)
+
+    for entry in (node_type, process_type):
+        if entry.count(LIKE_OPERATOR_CHARACTER) > 1:
+            raise ValueError('full type component `{}` contained more than one like-operator character'.format(entry))
+
+        if LIKE_OPERATOR_CHARACTER in entry and entry[-1] != LIKE_OPERATOR_CHARACTER:
+            raise ValueError('like-operator character in full type component `{}` is not at the end'.format(entry))
+
+    if LIKE_OPERATOR_CHARACTER in node_type:
+        filters['node_type'] = {'like': node_type}
+    else:
+        filters['node_type'] = node_type
+
+    if LIKE_OPERATOR_CHARACTER in process_type:
+        filters['process_type'] = {'like': process_type}
```

Comment:
Do we need/want to do something if the underscore `_` is in the process type? Should we escape by default or not?

Reply:
the underscore will certainly be there, for example for `CalcFunctionNode` whose process type includes the function name, which often contains underscores. So yes this should be escaped

Reply:
added escaping for sql like filters

Reply:
There is still a problem in the way escaping is done (final `%` is also escaped) - to be fixed